subsys/disk/disk_access_sdhc.c

```diff
@@ -0,0 +1,1001 @@
+#define SDHC_TOKEN_MULTI_WRITE 0xFC
+#define SDHC_TOKEN_STOP_TRAN 0xFD
+
+/* Data block responses */
+#define SDHC_RESPONSE_ACCEPTED 0x05
+#define SDHC_RESPONSE_CRC_ERR 0x0B
+#define SDHC_RESPONSE_WRITE_ERR 0x0E
+
+/* Clock speed used during initialisation */
+#define SDHC_INITIAL_SPEED 400000
+/* Clock speed used after initialisation */
+#define SDHC_SPEED 4000000
+
+#define SDHC_MIN_TRIES 20
+#define SDHC_RETRY_DELAY K_MSEC(20)
+/* Time to wait for the card to initialise */
+#define SDHC_INIT_TIMEOUT K_MSEC(5000)
+/* Time to wait for the card to respond or come ready */
+#define SDHC_READY_TIMEOUT K_MSEC(500)
+
+struct sdhc_data {
+	struct spi_config cfg;
+	struct device *cs;
+	u32_t pin;
+
+	u32_t sector_count;
+	u8_t status;
+	int trace_dir;
+};
+
+struct sdhc_retry {
+	u32_t end;
+	s16_t tries;
+	u16_t sleep;
+};
+
+struct sdhc_flag_map {
+	u8_t mask;
+	u8_t err;
+};
+
+DEVICE_DECLARE(sdhc_0);
+
+/* The SD protocol requires sending ones while reading but Zephyr
+ * defaults to writing zeros.
+ */
+static const u8_t sdhc_ones[] = {
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+};
+
+BUILD_ASSERT(sizeof(sdhc_ones) % SDHC_CSD_SIZE == 0);
+BUILD_ASSERT(SDHC_SECTOR_SIZE % sizeof(sdhc_ones) == 0);
+
+/* Maps R1 response flags to error codes */
+static const struct sdhc_flag_map sdhc_r1_flags[] = {
+	{SDHC_R1_PARAMETER, EFAULT},       {SDHC_R1_ADDRESS, EFAULT},
+	{SDHC_R1_ILLEGAL_COMMAND, EINVAL}, {SDHC_R1_COM_CRC, EILSEQ},
+	{SDHC_R1_ERASE_SEQ, EIO},	  {SDHC_R1_ERASE_RESET, EIO},
+	{SDHC_R1_IDLE, ECONNRESET},	{0, 0},
+};
+
+/* Maps disk status flags to error codes */
+static const struct sdhc_flag_map sdhc_disk_status_flags[] = {
+	{DISK_STATUS_UNINIT, ENODEV},
+	{DISK_STATUS_NOMEDIA, ENOENT},
+	{DISK_STATUS_WR_PROTECT, EROFS},
+	{0, 0},
+};
+
+/* Maps data block flags to error codes */
+static const struct sdhc_flag_map sdhc_data_response_flags[] = {
+	{SDHC_RESPONSE_WRITE_ERR, EIO},
+	{SDHC_RESPONSE_CRC_ERR, EILSEQ},
+	{SDHC_RESPONSE_ACCEPTED, 0},
+	/* Unrecognised value */
+	{0, EPROTO},
+};
+
+static u8_t sdhc_crc7(u8_t seed, const u8_t *src, size_t len)
+{
+	while (len--) {
+		u8_t ch = *src++;
+		u8_t d0 = ch & 1;
+
+		u8_t e = seed ^ (ch >> 1);
+
+		u8_t e345 = (e >> 3) & 7;
+		u8_t e01 = e & 3;
+		u8_t e6 = (e >> 6) & 1;
+		u8_t e2 = (e >> 2) & 1;
+
+		u8_t t = e345 ^ (e01 << 1) ^ d0;
+
+		u8_t t1 = t ^ (t << 3);
+		u8_t t2 = e6 ^ (e2 << 3) ^ ((e6 ^ e2) << 6);
+		u8_t t3 = e345 << 4;
+
+		seed = t1 ^ t2 ^ t3;
+	}
+
+	return seed;
+}
+
+static u8_t sdhc_crc7_finish(u8_t crc) { return (crc << 1) | 1; }
+
+static u16_t sdhc_crc16(u16_t seed, const u8_t *src, size_t len)
+{
```

Comment:
there is already a generic implementation of crc16, see include/crc16.h

Reply:
Ah, that's more epic.  The one in include/crc16 bitwise and 10x slower than this bytewise version (see #5098) and would drop the transfer rate by 2.3x [1].  My plan was to land this PR first, then pick up #5098, and then add support in drivers/crc for the hardware CRC unit on some devices (which is 4x faster again on the SAMD21).  WDYT?

[1] https://juju.nz/michaelh/post/2017/sdhc_speed/